File: src/AgentAcademy.Shared/Models/GoalCards.cs
using System.ComponentModel.DataAnnotations;

namespace AgentAcademy.Shared.Models;

/// <summary>
/// Snapshot of a goal card — the structured intent artifact an agent creates
/// before starting significant work. Captures task vs. intent analysis for
/// drift detection. Content is immutable after creation; only status transitions.
/// </summary>
public record GoalCard(
    string Id,
    string AgentId,
    string AgentName,
    string RoomId,
    string? TaskId,
    string TaskDescription,
    string Intent,
    string Divergence,
    string Steelman,
    string Strawman,
    GoalCardVerdict Verdict,
    string FreshEyes1,
    string FreshEyes2,
    string FreshEyes3,
    int PromptVersion,
    GoalCardStatus Status,
    DateTime CreatedAt,
    DateTime UpdatedAt
);

/// <summary>
/// Request to create a goal card. All content fields are required.
/// </summary>
public record CreateGoalCardRequest(
    [Required, StringLength(2000, MinimumLength = 10)]
    string TaskDescription,

    [Required, StringLength(2000, MinimumLength = 10)]
    string Intent,

    [Required, StringLength(500, MinimumLength = 5)]
    string Divergence,

    [Required, StringLength(2000, MinimumLength = 20)]
    string Steelman,

    [Required, StringLength(2000, MinimumLength = 20)]
    string Strawman,

    [Required]
    GoalCardVerdict Verdict,

    [Required, StringLength(1000, MinimumLength = 10)]
    string FreshEyes1,

    [Required, StringLength(1000, MinimumLength = 10)]
    string FreshEyes2,

    [Required, StringLength(1000, MinimumLength = 10)]
    string FreshEyes3,

    string? TaskId = null
);

/// <summary>
/// Request to update the status of a goal card.
/// </summary>
public record UpdateGoalCardStatusRequest(
    [Required] GoalCardStatus Status
);

File: src/agent-academy-client/src/api/goalCards.ts
import { apiUrl, request } from "./core";

// --- Types ---

export type GoalCardVerdict = "Proceed" | "ProceedWithCaveat" | "Challenge";
export type GoalCardStatus = "Active" | "Completed" | "Challenged" | "Abandoned";

export interface GoalCard {
  id: string;
  agentId: string;
  agentName: string;
  roomId: string;
  taskId: string | null;
  taskDescription: string;
  intent: string;
  divergence: string;
  steelman: string;
  strawman: string;
  verdict: GoalCardVerdict;
  freshEyes1: string;
  freshEyes2: string;
  freshEyes3: string;
  promptVersion: number;
  status: GoalCardStatus;
  createdAt: string;
  updatedAt: string;
}

export interface GoalCardFilters {
  roomId?: string;
  agentId?: string;
  taskId?: string;
  status?: GoalCardStatus;
  verdict?: GoalCardVerdict;
}

// --- API functions ---

export function getGoalCards(filters?: GoalCardFilters): Promise<GoalCard[]> {
  const qs = new URLSearchParams();
  if (filters?.roomId) qs.set("roomId", filters.roomId);
  if (filters?.agentId) qs.set("agentId", filters.agentId);
  if (filters?.taskId) qs.set("taskId", filters.taskId);
  if (filters?.status) qs.set("status", filters.status);
  if (filters?.verdict) qs.set("verdict", filters.verdict);
  const query = qs.toString();
  return request<GoalCard[]>(apiUrl(`/api/goal-cards${query ? `?${query}` : ""}`));
}

export function getGoalCard(id: string): Promise<GoalCard> {
  return request<GoalCard>(apiUrl(`/api/goal-cards/${encodeURIComponent(id)}`));
}

export function updateGoalCardStatus(id: string, status: GoalCardStatus): Promise<GoalCard> {
  return request<GoalCard>(apiUrl(`/api/goal-cards/${encodeURIComponent(id)}/status`), {
    method: "PATCH",
    body: JSON.stringify({ status }),
  });
}
